### PR TITLE
Add Deterministic Miller-Rabin Primality Test

### PR DESCRIPTION
**Pull request by raklaptudirm**
### Tasks
- [x] Added deterministic version of Miller-Rabin test for valid `int64` numbers. Note the algorithm can still fail due to overflow while modular exponentiation.
- [x] Refactor the existing code.
- [x] Comment code and fix some old comments.

### CODE REVIEW

File: math/prime/millerrabinprimalitytest.go
Comment (tjgurwara99) [inline math/prime/millerrabinprimalitytest.go:33]:
```suggestion
func isTrivial(num int64) (bool, bool) {
```

This function is very confusing in its current design - why do we care whether this is trivial or not? We should only care about whether this is a prime or not.

Furthermore, you are using named return types - there is no readability benefit for them usually (in fact I think you're function signature is not right for its purpose). The main benefit of using named return types is pre-allocation (a compile time optimisation) but you are neither assigning any values to these or using them in any specific way making this redundant.

File: math/prime/millerrabinprimalitytest_test.go
Comment (tjgurwara99) [inline math/prime/millerrabinprimalitytest_test.go:31]: separate out the tests - they should be independent

File: math/prime/millerrabinprimalitytest_test.go
Comment (tjgurwara99) [inline math/prime/millerrabinprimalitytest_test.go:47]: Separate out benchmarks please - they skew the actual results.

**Review by tjgurwara99 — CHANGES_REQUESTED** [review]

File: math/prime/millerrabinprimalitytest.go
Comment (raklaptudirm) [reply math/prime/millerrabinprimalitytest.go:24]: Wikipedia uses the `s` notation for this, so I changed it since we link to Wikipedia as a reference.

File: math/prime/millerrabinprimalitytest.go
Comment (raklaptudirm) [reply math/prime/millerrabinprimalitytest.go:33]: > This function is very confusing in its current design - why do we care whether this is trivial or not? We should only care about whether this is a prime or not.

`isTrivial` checks if a number's primality can be trivially determined. For example, all multiples of 2, other than 2 itself, are not primes, so we do not have to run the Miller-Rabin test in that case (also you can't write even numbers as `2^s * d + 1`). Also, the primality of 2 and 3 can be determined trivially as prime. If we can determine the primality without running the tests, the function will return true and the primality of the number. Otherwise, we have to run the test since we can't determine it's primality, so it returns false and false in that case.

> Furthermore, you are using named return types - there is no readability benefit for them usually (in fact I think you're function signature is not right for its purpose). The main benefit of using named return types is pre-allocation (a compile time optimisation) but you are neither assigning any values to these or using them in any specific way making this redundant.

Named returned types appear in the docs, so they are useful as documentation.

File: math/prime/millerrabinprimalitytest_test.go
Comment (raklaptudirm) [reply math/prime/millerrabinprimalitytest_test.go:54]: Done.

File: math/prime/millerrabinprimalitytest_test.go
Comment (raklaptudirm) [reply math/prime/millerrabinprimalitytest_test.go:36]: Done.

**Comment by zhrgithub** [comment]
done


| |
zhr
|
|
***@***.***
|
签名由网易邮箱大师定制


在2022年03月28日 13:15，Rak ***@***.***> 写道：

@raklaptudirm commented on this pull request.

In math/prime/millerrabinprimalitytest_test.go:

> @@ -23,7 +18,17 @@ func TestMillerRabinTest(t *testing.T) {
 	}
 	for _, test := range tests {
 		t.Run(test.name, func(t *testing.T) {
-			output, err := MillerRabinTest(test.input, test.rounds)
+			output, err := MillerRabinProbabilistic(test.input, test.rounds)
+			if err != test.err {
+				t.Errorf("For input: %d, unexpected error: %v, expected error: %v", test.input, err, test.err)
+			}
+			if output != test.expected {
+				t.Errorf("For input: %d, expected: %v", test.input, output)
+			}
+		})
+
+		t.Run(test.name, func(t *testing.T) {
+			output, err := MillerRabinDeterministic(test.input)


Done.

**Review by tjgurwara99 — CHANGES_REQUESTED** [review]

**Comment by raklaptudirm** [comment]
I am a modular approach guy so since we were repeating those tests, I made a function out of them.

**Review by siriak — APPROVED** [review]

**Comment by tjgurwara99** [comment]
> I am a modular approach guy so since we were repeating those tests, I made a function out of them.

Where?

**Comment by raklaptudirm** [comment]
I was talking about the `isTrivial` function. Since it is required in both the deterministic and the probabilistic tests.

**Comment by tjgurwara99** [comment]
> I was talking about the isTrivial function.

I disagree with that statement actually. Your approach is concise, not clean imo (there is a huge difference between these two). I think you might disagree with this but I really think you have a different understanding of what Modularity is.

Modular in the context of software engineering is vague to begin with, so before that I would like to define what I mean by it. Googling the terms gives us a few definitions of Modularity - one of which is here which corresponds to my understanding `In software engineering, modularity refers to the extent to which a software/Web application may be divided into smaller modules.` This definition was taken from [here](https://www.techopedia.com/definition/24772/modularity).

Now, with this in mind, I would argue that your code is less modular than the suggestion I made. The division of the functions into two functions distributed with respect their responsibility, then I think you will agree. With that said however, there are still things that we should consider that goes hand in hand with this principle of modularity. Its called the single responsibility principle, the idea is quite straight forward but very powerful when it comes to software engineering - I really think you should study it. It's the corner stone of SOLID principles defined by Uncle Bob.

Anyways, I leave it up to you whether you want to change that or not, I think you should.

**Comment by raklaptudirm** [comment]
Well, we have 3 maintainers for a reason :). @siriak  What do you think?

File: math/prime/millerrabinprimalitytest.go
Comment (siriak) [reply math/prime/millerrabinprimalitytest.go:33]: I'd argue that the SRP is not violated here because the purpose of this function is not to determine primality, but only **try** to do that. And it returns the result and the flag if it has succeeded or not. This is a common pattern in other languages for operations that can fail (e.g. parsing a string to a number)

**Review by siriak — APPROVED** [review]
Overall I like this design, the `isTrivial` function could be split further as suggested, but I think it's fine as-is because the operations performed are trivial.

File: math/prime/millerrabinprimalitytest.go
Comment (tjgurwara99) [reply math/prime/millerrabinprimalitytest.go:33]: Fair enough, if you disagree then majority rules. If you're talking about the pattern of result and a flag, then it still doesn't follow idiomatic Go calling structure (`res, ok` rather than `ok, res`).

Anyways, I feel like this is a redundant thing since majority agrees but I still think SRP is violated. The `ok` flag is valid pattern in Go as well but trivial flag is by its nature testing something else rather than testing whether this is a prime (just thinking of the naming of the function itself points that out) so the pattern doesn't really apply here - note that there is separate condition testing for completely separate thing - which I think violates the principle but... I concede.

**Review by tjgurwara99 — DISMISSED** [review]
Majority agrees with this change so here's my approval 😄

File: math/prime/millerrabinprimalitytest.go
Comment (raklaptudirm) [reply math/prime/millerrabinprimalitytest.go:33]: Changing it to follow the comma ok idiom is a good idea, will make the changes.

Comment (raklaptudirm) [reply math/prime/millerrabinprimalitytest.go]: @tjgurwara99 [Here](https://github.com/golovers/effective-go#named-result-parameters) is an excerpt from a document I quite like, which explains my usage of named returns.

Comment (raklaptudirm) [reply math/prime/millerrabinprimalitytest.go]: Also, I made the code more concise using named returns.

Comment (tjgurwara99) [reply math/prime/millerrabinprimalitytest.go]: > @tjgurwara99 [Here](https://github.com/golovers/effective-go#named-result-parameters) is an excerpt from a document I quite like, which explains my usage of named returns.

Will take a look 😄

**Comment by raklaptudirm** [comment]
@tjgurwara99 @siriak ping

**Review by tjgurwara99 — APPROVED** [review]

**Review by siriak — APPROVED** [review]

**Pull request merged**
